activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb

```diff
@@ -5,15 +5,30 @@ module ConnectionAdapters
     module PostgreSQL
       class SchemaCreation < SchemaCreation # :nodoc:
         private
-          delegate :quoted_include_columns_for_index, to: :@conn
+          delegate :quoted_include_columns_for_index, :create_enum, to: :@conn
+
+          def visit_TableDefinition(o)
+            create_enums(o.columns)
```

Comment:
It looks like `create_enums` immediately executes the query, I'm not sure that's an expected behavior of a visitor. I assume a visitor should visit all nodes and build a SQL query which will be executed later. Even though none of the tests seem to be failing, should we consider splitting `create_enum` into `build_enum_sql` + `execute_query` so we can reuse `build_enum_sql` here and prepend the enum statement. 

Even though I can't point where the issue is exactly, I feel like the resulting sql built by a visitor should include enum statements and none of the `visit_` methods should execute any query, just build sql

Reply:
I attempted to do so originally however, in `add_column_options!` which gets executed a bit later (in `visit_ColumnDefinition`), `lookup_cast_type` is called through [`quote_default_expression_for_column_definition`](https://github.com/Shopify/rails/blob/ae63129bd932d7e2997076012d8bddb24de0c8a2/activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb#L151) which executes a query looking for the type. 

Since the enum type hasn't been created, this line would error (eg. `ActiveRecord::StatementInvalid: PG::UndefinedObject: ERROR:  type "mood" does not exist`). So I doubt that we can delay the execution. I think it's acceptable since we were sending separate queries to create enums before.